### PR TITLE
[SPARK-33259][SS] Disable streaming query with possible correctness issue by default

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to disable the streaming query with possible correctness issue in chained stateful operators. The behavior can be controlled by a SQL config, so if users understand the risk and still want to run the query, they can disable the check.

### Why are the changes needed?

The possible correctness in chained stateful operators in streaming query is not straightforward for users. From users perspective, it will be considered as a Spark bug. It is also possible the worse case, users are not aware of the correctness issue and use wrong results.

A better approach should be to disable such queries and let users choose to run the query if they understand there is such risk, instead of implicitly running the query and let users to find out correctness issue by themselves and report this known to Spark community. 

### Does this PR introduce _any_ user-facing change?

Yes. Streaming query with possible correctness issue will be blocked to run, except for users explicitly disable the SQL config.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @dongjoon-hyun @HeartSaVioR

**Comment by dongjoon-hyun** [comment]
Thank you so much, @viirya !

**Comment by HeartSaVioR** [comment]
I'd rather try to get some sort of consensus via initiating the discussion around dev@ mailing list. I see Spark community does everything in the PR (even that requires some sort of consensus) which I don't think it's ideal and the result is from consensus among the narrow group.

To explain why I made it just logging instead of failing the query - I tried to get consensus around how to deal with this before:

* https://lists.apache.org/thread.html/cc6489a19316e7382661d305fabd8c21915e5faf6a928b4869ac2b4a@%3Cdev.spark.apache.org%3E
* #24890

I failed to get any voice except @gaborgsomogyi in #24890 and the approach wasn't radical so I did it.

This change may break some query which may work if end users are super careful and know in details. I don't expect majority of end users could be, but just hypothetically thinking. I'm OK to disable such query, but I'm not 100% sure everyone is on the same page. (Someone might concern.)

**Comment by HeartSaVioR** [comment]
cc. @tdas @zsxwing @jose-torres @gaborgsomogyi @xuanyuanking

**Comment by dongjoon-hyun** [comment]
cc @dbtsai , @gatorsmile since this is a correctness issue.

**Comment by xuanyuanking** [comment]
I think it's ok to change the original param to a SQL config for end users.
```
This change may break some query which may work if end users are super careful and know in details and go ahead.
```
+1 for this concern. So how about change the default value to `false`?

```
I failed to get any voice except @gaborgsomogyi in #24890 and the approach wasn't radical so I did it.
```
Actually, I reviewed that PR after merging :joy:, thanks for the excellent doc! 
@viirya qq: Do we have the real cases on enabling this config without correctness issues? It would be great to keep updating the document by providing demo cases and specific usage of this config.

**Comment by viirya** [comment]
> I think it's ok to change the original param to a SQL config for end users.
> 
> ```
> This change may break some query which may work if end users are super careful and know in details and go ahead.
> ```
> 
> +1 for this concern. So how about change the default value to `false`?
> 

I believe this is not the first change that may break some queries. We did some similar. For such changes, we provided some configs so users still can keep with legacy behavior if they want. This change basically follows this approach.

This involves correctness and may not be aware by users. Users need to be very careful to avoid the issue. I think we should provide a baseline which is definitely correct, and provide an option (the config) for users to run with correctness risk.

> @viirya qq: Do we have the real cases on enabling this config without correctness issues? It would be great to keep updating the document by providing demo cases and specific usage of this config.

For outer join or aggregation, I think the risk of correctness is pretty high. `FlatMapGroupsWithState`, I am not sure, but I think it is possible to not emit late rows in the state function, maybe @HeartSaVioR has some real cases?

**Comment by HeartSaVioR** [comment]
No I don't have real case for knowing and taking the risk. Probably I could create some query which could evade the issue, but I agree that's more likely in theory and not real case.

Saying again I don't object the change. If you look back my proposal then you'll find blocking the query is also one of options in my proposal. My point was that such change warrants the discussion, ideally in dev@ mailing list instead of PR.

**Comment by viirya** [comment]
Hmm, based on what I saw, it seems to me the discussion on dev@ mailing list is not so active, but I'm okay to drop some words in dev@ mailing list and see if we can get some feedback.

**Comment by HeartSaVioR** [comment]
That's the chicken and egg problem, you know. dev@ list is not so active because all the important discussions aren't passing through the dev list, which is I think bad in perspective of "community over code".

**Comment by viirya** [comment]
Post to dev mailing list: http://apache-spark-developers-list.1001551.n3.nabble.com/DISCUSS-Disable-streaming-query-with-possible-correctness-issue-by-default-td30380.html

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (tgravescs) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1389]: this should have more information about the correctness issue or point to somewhere that does so users can properly make a decision.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
cc @tdas , @zsxwing , @cloud-fan , @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1389]: Added more info. Thanks.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by xuanyuanking** [comment]
LGTM for the code changes. Let's also mention the behavior change in ss-migration-guide.md

**Comment by viirya** [comment]
> LGTM for the code changes. Let's also mention the behavior change in ss-migration-guide.md

Thanks @xuanyuanking. I will put it to the migration guide.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1386]: And once we change we also need to change UnsupportedOperationChecker as well.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala:1329]: If we need to turn off the config to run the test, I think the test needs to be reconsidered - we may give misleading and discouraged patterns from tests.

That's not necessarily to be addressed in this PR altogether (follow-up or a new JIRA issue), as removing tests have been always a concern and that leads another kind of discussion.

**Review by HeartSaVioR — COMMENTED** [review]
Thanks for going through discussion thread. I guess we have enough number of responses so you can conclude the mail thread in dev mailing list and be back here.

**Comment by dongjoon-hyun** [comment]
Shall we finalize and merge this PR by address @HeartSaVioR 's comments?

**Comment by HeartSaVioR** [comment]
> Let's also mention the behavior change in ss-migration-guide.md

Let's make sure this review comment is also addressed as well. I just skipped mentioning it as it's already commented.

**Comment by viirya** [comment]
> Shall we finalize and merge this PR by address @HeartSaVioR 's comments, @viirya ?

Yeah, I will address these comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1386]: For the naming, cc @cloud-fan .

**Review by tgravescs — APPROVED** [review]
I'm not super picky on the config name, we were trying to get rid of extra .xxx. and nothing else uses .statefulOperator. but changing to combine with checkCorrectness seems very long so seems fine to me

**Comment by viirya** [comment]
retest this please

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @viirya and all!
Merged to master for Apache Spark 3.1.

**Comment by HeartSaVioR** [comment]
(I hope the PR can wait for reviewers' approvals who left valid review comments. I understand the PR can be merged "technically", but doesn't seem to be a good practice. If in doubt w.r.t reviewers being inactive, we can at least wait for a day based on timezone difference.)

**Comment by dongjoon-hyun** [comment]
Sorry, but could you be more specific? It seems that I missed some comments.

**Comment by HeartSaVioR** [comment]
There's no indication I have produced all review comments (while actually I produced all review comments), and review comment author would be the right one to verify whether the review comments are addressed as suggested. I planned to review the new change again today but it's already merged.

I see we tend to ask reviewers to see whether they're happy with the last change before merging. At least we "were", and I tend to ensure that, unless reviewers are not responded for days. I'd hope we ensure everyone is happy with the change before merging, otherwise discuss/debate if necessary.

**Comment by dongjoon-hyun** [comment]
First of all, there is no domain owner in Apache Spark. There is no way for the other people know your mind (or your review plan). Given that, your claim sounds like you are saying you are not happy because someone-else merged it without your approval in `Structure Streaming` domain. Please make it sure that I didn't ignore your veto or negative comment in this PR. I'm sure about this PR, too.

**Comment by dongjoon-hyun** [comment]
Please leave a note if you think you should approve this PR before merging.

**Comment by HeartSaVioR** [comment]
No that's not also my point. I don't claim about the domain owner (but in practice I see there's implicit domain owner). You can merge the PR in SS area without me if I didn't review.

The concern is that I was the one of active reviewers and you're just assuming my review comments were addressed without my confirmation. Both my and @xuanyuanking's comments. I don't require others to read my mind, and anyone can't read my mind, so there should be the process of "confirmation". That would just bring delay of merging a day or so, wouldn't hurt anything.

**Comment by dongjoon-hyun** [comment]
That's my first question, @HeartSaVioR . Please be specific what is missed from your comment and @xuanyuanking ?
> Both my and @xuanyuanking's comments. 

This PR was created based on my request on JIRA and has been open for two weeks. And, there was no fundamental code change for two weeks. Only building consensus and adding documents and revising configuration. I don't see any technical reviews from you.

**Comment by dongjoon-hyun** [comment]
Specifically, you and I have different idea on this. From my side, this has been here for two weeks to build consensus. From your side, it's one day, right?
> That would just bring delay of merging a day or so

**Comment by HeartSaVioR** [comment]
What do you say? I don't say you're missing the comment. I say the confirmation is better to be done by the reviewers who review, not from some other one. The confirmation would only take a day or so (due to the timezone) if the reviewers are quite active. Even you can say "I'll merge in tomorrow if there's no further feedback." to force the delay to a day.

You're emphasizing 2 weeks but I can enumerate the PRs which waits for a half of year or even years. The code change is quite small (because the detection logic was there), but given it's breaking backward compatibility, it's not odd request to make a consensus. If the default is false probably we rather concern less.

**Comment by dongjoon-hyun** [comment]
@HeartSaVioR . When a committer approves a PR, it means it looks good to him and it means the PR can be merged technically at any time. If you didn't ignore it, you can notice that I did it two days ago.
> Even you can say "I'll merge in tomorrow if there's no further feedback." to force the delay to a day.

I'd be happy if you respect my vote or approval as I did for yours. Again, I want to emphasize that I tried to check your and other reviews' comments here. That's the reason why I'm surprised at your re-action here.

**Comment by HyukjinKwon** [comment]
It's just that both points are valid ... It would have been best for a couple of days given that we had a sort of long discussions/reviews but I don't see any serious problem about the fact that it was merged because people can continue reviewing after this PR was merged. It will be reverted at the worst case.

**Comment by HeartSaVioR** [comment]
I'd rather avoid the chance of "post-review" whenever possible, but I'd admit everyone has different thoughts. I'm OK with it, and if that's considered here (and no one would be disappointed from post-review), I was clearly wrong. My bad.

I hope we prefer to say about "good practice" of doing something instead of saying "whether someone is the right to do something". Everything is technically possible. e.g. I don't think Spark has some sort of BYLAWS which prevents committers to self +1 and push, but I don't leverage it even I still have PRs sitting around years which got LGTM from contributors. It's just because it's a bad practice.

Agree to disagree about the different understanding of "details" about good practice of reviewing is OK (because I may be wrong), but I'd be unhappy if we keep saying about the "right". This says I should always start with explicitly vote -1 whenever I start reviewing, which is the only way about the claim of "my right" of reviewer, but no one would do this as we have a belief and huge negative effect on putting veto. And sure I don't change my reviewing practice to start from -1. (That's gonna be just funny.)

**Comment by HyukjinKwon** [comment]
It might have been better to wait a bit more in case other people have more comments, sure. For a post-review, I think it's fine. The discussions and changes can still be seen as are. It's just a little bit inconvenient to push some more changes or review, though.

This PR had multiple approvals and reviews from committers and seems fine to merge. It's okay, maybe good.

**Comment by dongjoon-hyun** [comment]
@HeartSaVioR . It seems that you tend to frame the Apache Spark community in negative ways. For your examples, it clearly looks like a straw-man attack. All of us didn't claim such funny things.

Putting them aside, I want to say that I have been appreciate your contributions really, especially in this area, `Structural Streaming` because this area really needs lots of more care from the community. To be constructive, it would be great if 
- You can let us know what was your un-address comments in this PR  (https://github.com/apache/spark/pull/30210#issuecomment-726413507)
- You can do review after commits. Please do it now instead of stopping like this. Apache Spark community is widely open for that and I am also happy to accept any comments at any timeframe.
- You can reply on the corresponding consensus email threads about this PR (which we already did based on your advice on JIRA and here) if you changed your mind. We are open for that, too.
- You can raise another email thread about your specific Spark practice requests like you did before. I was also happy about your issue raising before and will be at this time again.

I believe both of us want to collaborate in a productive way in this area, @HeartSaVioR .

**Comment by HeartSaVioR** [comment]
I was writing a wall of text and Chrome happily (?) killed itself. Rewriting one.

What I really asked you to do is exactly this. The practice is also happened for the first time I made a first contribution on Apache Spark. If you want I can find that too.

https://github.com/apache/spark/pull/30139#issuecomment-727103124

I'm really unhappy now you're saying I'm framing Apache Spark community to be negative ways. I was never saying about the community here. Let me quote my comment again, where's the strong voice like the attacking or making things being negative, and mentioning any sort of thing representing community?

> (I hope the PR can wait for reviewers' approvals who left valid review comments. I understand the PR can be merged "technically", but doesn't seem to be a good practice. If in doubt w.r.t reviewers being inactive, we can at least wait for a day based on timezone difference.)

This comment is exactly explaining the same thing what I link for reference. Could you find any difference? Even I didn't say that strongly.

A PMC member is just a "one of" PMC members. Before you make consensus along with PMC members and construct a BYLAWS, you can't frame me as one guy complaining to Spark community. The practice is not a thing "everyone" has a belief, so attacking the practice shouldn't be interpreted like so. Makes sense?

For sure, I said about -1 as a sort of joke, but what you requested to me to block merger to merge at any time is "technically" the same. That is just a "representational" difference, and for me it's also not welcomed one if I have to always mention merger to stop doing before I approve. And this suggestion is your own, not from the consensus.

But in overall, I have to admit I really did one bad thing, trying to educate someone which is not constructed like BYLAW in any way, regardless of the intention. I should avoid doing that and raise discussion in dev mailing thread like I have been doing before. My apologize.

**Comment by HeartSaVioR** [comment]
And I also admit I have different voice on post-review. I agree post-review would open the possibility for reviewers to review later who weren't active during the review period before merging, but IMHO it's over-use of post-review if it's being used to provide merger rationalization to cut the review period at any time, feel other reviewers to be **disregarded**.

**Comment by dongjoon-hyun** [comment]
Please remember that I also replied politely at your first comment like this. I apologized first and tried to understand your request and to fix it immiediately because I thought I missed your some comments.
> Sorry, but could you be more specific? It seems that I missed some comments.

But, it isn't true, is it? You didn't make any signal here. Suddenly, you were upset that your approval is not asked explicitly by me.

After than, you did ignore my approval step and suggested me to have to say `I'll merge in tomorrow if there's no further feedback.`. That's the way I feel that you disregarded me. The rest of discussion looks like the same pattern to me.

**Comment by dongjoon-hyun** [comment]
You are saying that I had taken your right away repeatedly. In this PR's context, it doesn't make sense to me either.

**Comment by HeartSaVioR** [comment]
I was feeling odd and became feeling upset because my intention wasn't change from the first comment and the intention was disregarded (at least that's what I felt like) for a couple of early replies, even I emphasized whether you checked my comments were addressed or not is not the point of concern here. 

The replies were simply based on whether there was any fault or not, and in addition, trying to interpret my action on your view and framed me. I never said my approval is required because I'm domain expert of SS. I mentioned twos including me because twos are only the reviewers who left "actual" review comments, not saying we're dealing with SS hence we are gatekeeper. This is unchanged from the first comment and I never said about SS in my comments. Same happened for the statement I'm framing the Apache Spark community be the negative. I'm going to be a bad guy based on what I even didn't mention. How I can't be upset?

Can we make sure we carefully read the comment and try not to interpret and just ask if we're unsure about the intention?

Other than that I feel this is a waste of time. I have been trying to explain the committer's power to say "technically correct" is not always representing "we are doing good", but I also admit it's super hard to reconsider the practice we have been using for a while, and I feel the air that we're reluctant to even think about it. At least I shouldn't deal with individual, as it's against my voice as well and that feels someone to be educated. My apologize again.

**Comment by dongjoon-hyun** [comment]
Can we talk over the phone? It would be better for us.

**Comment by HeartSaVioR** [comment]
Sorry to all for all noises. Please disregard all conversation. I'll remove them now.

**Comment by HeartSaVioR** [comment]
I just initiated the discussion on dev@ mailing list which I should have been done instead.
https://lists.apache.org/thread.html/r30069e17f59e8d29267ae296d56840970905476019023f20164ee5a3%40%3Cdev.spark.apache.org%3E

My apologize to make noises and feel anyone unhappy.

**Pull request closed without merging**
